Add docs for new validation features in .NET 11 Preview 5 (#37235)
dotnet/AspNetCore.Docs: aspnetcore/blazor/forms/validation.md

In the following example, a custom validator component checks a username against a remote endpoint when the form is submitted:

```razor
@implements IDisposable
@inject HttpClient Http

@code {
    [CascadingParameter]
    private EditContext? CurrentEditContext { get; set; }

    [Parameter, EditorRequired]
    public RegistrationModel Model { get; set; } = default!;

    private ValidationMessageStore? _messages;

    protected override void OnInitialized()
    {
        ArgumentNullException.ThrowIfNull(CurrentEditContext);
        _messages = new ValidationMessageStore(CurrentEditContext);
        CurrentEditContext.OnValidationRequestedAsync += ValidateUsernameAsync;
    }

    private async Task ValidateUsernameAsync(
        object? sender, ValidationRequestedEventArgs e)
    {
        var field = CurrentEditContext!.Field(nameof(Model.Username));
        _messages!.Clear(field);

        var available = await Http.GetFromJsonAsync<bool>(
            $"api/usernames/available?value={Uri.EscapeDataString(Model.Username)}",

>             e.CancellationToken);
> 
>         if (!available)
>         {
>             _messages.Add(field, "The username is already taken.");
>         }
>     }

Author: ```suggestion
        if (!available)
        {
            _messages.Add(field, "The username is already taken.");
        }
        
        CurrentEditContext!.NotifyValidationStateChanged();
    }
```

```csharp
private CancellationTokenSource? _usernameCts;

protected override void OnInitialized()
{
    ArgumentNullException.ThrowIfNull(CurrentEditContext);
    _messages = new ValidationMessageStore(CurrentEditContext);
    CurrentEditContext.OnValidationRequestedAsync += ValidateUsernameAsync;
    CurrentEditContext.OnFieldChanged += OnFieldChanged;
}

private void OnFieldChanged(object? sender, FieldChangedEventArgs e)
{
    if (e.FieldIdentifier.FieldName != nameof(RegistrationModel.Username))
    {
        return;
    }

    _usernameCts?.Cancel();
    _usernameCts = new CancellationTokenSource();
    var token = _usernameCts.Token;

    var task = CheckAsync(e.FieldIdentifier, token);
    CurrentEditContext!.AddValidationTask(e.FieldIdentifier, task);
}

private async Task CheckAsync(FieldIdentifier field, CancellationToken token)
{
    _messages!.Clear(field);

    var available = await Http.GetFromJsonAsync<bool>(
        $"api/usernames/available?value={Uri.EscapeDataString(Model.Username)}",

>         token);
> 
>     if (!available)
>     {
>         _messages.Add(field, "The username is already taken.");
>     }
> }

Author: ```suggestion
    if (!available)
    {
        _messages.Add(field, "The username is already taken.");
    }
    
    CurrentEditContext!.NotifyValidationStateChanged();
}
```

* Any manual validation that you want to perform in the `HandleValidationRequested` method assigned to the form's <xref:Microsoft.AspNetCore.Components.Forms.EditContext.OnValidationRequested%2A> event executes when the user selects the form's **`Update`** button. In the existing code of the `Starship8` component example, the user must select either or both of the checkboxes to validate the form.
* The form doesn't process the `Submit` method until both the data annotations and manual validation pass.

:::moniker range=">= aspnetcore-11.0"

## Asynchronous validation

<xref:Microsoft.AspNetCore.Components.Forms.EditContext> exposes an asynchronous validation pipeline that custom validator components and custom submit handlers can use to run validation work that performs I/O, such as calling a server endpoint to check a value's uniqueness. The pipeline is built around three additions to <xref:Microsoft.AspNetCore.Components.Forms.EditContext>:

> * <xref:Microsoft.AspNetCore.Components.Forms.EditContext.ValidateAsync%2A>: an asynchronous counterpart to <xref:Microsoft.AspNetCore.Components.Forms.EditContext.Validate%2A> that awaits any registered async work and accepts a <xref:System.Threading.CancellationToken>.

Reviewer: ```suggestion
<!-- UPDATE 11.0 - API Browser cross-link

<xref:Microsoft.AspNetCore.Components.Forms.EditContext.ValidateAsync%2A>

-->

* `Microsoft.AspNetCore.Components.Forms.EditContext.ValidateAsync`: an asynchronous counterpart to <xref:Microsoft.AspNetCore.Components.Forms.EditContext.Validate%2A> that awaits any registered async work and accepts a <xref:System.Threading.CancellationToken>.
```

}
else if (EditContext.IsValidationFaulted(() => Model.Username))
{
    <span class="validation-faulted" aria-live="polite">
        Validation could not be completed.
    </span>
}
```

The form-level parameterless overloads return `true` when any field is currently pending or faulted. A common use is disabling the submit button while validation is in flight:

```razor
<button type="submit" disabled="@EditContext.IsValidationPending()">
    Register
</button>
```

<xref:Microsoft.AspNetCore.Components.Forms.InputBase%601> automatically adds the `pending` and `faulted` CSS classes to its rendered element while the bound field is in the corresponding state, in addition to the existing `modified` / `valid` / `invalid` classes. The classes compose, so unmodified pending styling and modified pending styling can be targeted independently:

```css
.pending {
    background-image: url('spinner.gif');
    background-repeat: no-repeat;
    background-position: right center;
}

.modified.pending {
    border-color: lightblue;
}

.modified.faulted {
    border-color: orange;
}
```

### Calling `ValidateAsync` from a custom submit handler

> When a form uses <xref:Microsoft.AspNetCore.Components.Forms.EditForm.OnSubmit> instead of <xref:Microsoft.AspNetCore.Components.Forms.EditForm.OnValidSubmit%2A>, call <xref:Microsoft.AspNetCore.Components.Forms.EditContext.ValidateAsync%2A> from the handler to await any registered async work before deciding whether to proceed:

Reviewer: ```suggestion
<!-- UPDATE 11.0 - API Browser cross-link 

<xref:Microsoft.AspNetCore.Components.Forms.EditContext.ValidateAsync%2A>

-->

When a form uses <xref:Microsoft.AspNetCore.Components.Forms.EditForm.OnSubmit> instead of <xref:Microsoft.AspNetCore.Components.Forms.EditForm.OnValidSubmit%2A>, call `Microsoft.AspNetCore.Components.Forms.EditContext.ValidateAsync` from the handler to await any registered async work before deciding whether to proceed:
```

```razor
<EditForm EditContext="_editContext" OnSubmit="HandleSubmitAsync">
    <UsernameUniquenessValidator Model="Model" />
    <InputText @bind-Value="Model.Username" />
    <button type="submit">Register</button>
</EditForm>

@code {
    private EditContext _editContext = default!;

    protected override void OnInitialized() =>
        _editContext = new EditContext(Model);

    private async Task HandleSubmitAsync()
    {
        if (await _editContext.ValidateAsync(CancellationToken.None))
        {
            await RegisterAsync();
        }
    }
}
```

> The synchronous <xref:Microsoft.AspNetCore.Components.Forms.EditContext.Validate%2A> method continues to work for forms that only have synchronous validators. When a registered async handler returns an incomplete <xref:System.Threading.Tasks.Task>, calling <xref:Microsoft.AspNetCore.Components.Forms.EditContext.Validate%2A> throws an <xref:System.InvalidOperationException> directing the caller to use <xref:Microsoft.AspNetCore.Components.Forms.EditContext.ValidateAsync%2A> instead.

Reviewer: ```suggestion
<!-- UPDATE 11.0 - API Browser cross-link

<xref:Microsoft.AspNetCore.Components.Forms.EditContext.ValidateAsync%2A>

-->

The synchronous <xref:Microsoft.AspNetCore.Components.Forms.EditContext.Validate%2A> method continues to work for forms that only have synchronous validators. When a registered async handler returns an incomplete <xref:System.Threading.Tasks.Task>, calling <xref:Microsoft.AspNetCore.Components.Forms.EditContext.Validate%2A> throws an <xref:System.InvalidOperationException> directing the caller to use `Microsoft.AspNetCore.Components.Forms.EditContext.ValidateAsync` instead.
```
